aws_lambda_builders/workflows/java_maven/DESIGN.md

```diff
@@ -0,0 +1,124 @@
+# Java - Maven Lambda Builder
+
+## Scope
+
+This package enables the creation of a Lambda deployment package for Java
+projects managed using the Maven build tool.
+
+## Challenges
+
+- Java Version compatibility mentioned in the [Gradle Lambda Builder] Design doc.
+
+- Building Multimodule project
+
+Here `ProjectA` is a a single lambda function, and `ProjectB` is a multimodule
+project where sub modules `lambda1` and `lambda2` are each a lambda
+function. In addition, suppose that `ProjectB/lambda1` has a dependency on its
+sibling module `ProjectB/common`.
+
+**Single Project A**
+```
+ProjectA
+├── pom.xml
+├── src
+└── template.yaml
+```
+
+**Multmodule Project B**
+```
+ProjectB
+├── common
+│   └── pom.xml
+├── lambda1
+│   └── pom.xml
+├── lambda2
+│   └── pom.xml
+├── pom.xml
+└── template.yaml
+```
+
+Building Project A is relatively simple since we just need to issue `mvn
+package` and place the built classes and dependency jars within the artifact directory.
+
+Building `ProjectB/lambda1` requires maven to build `lambda1` module from
+the root pom directory and use `--also-make` option to build necessary dependencies 
+(`ProjectB/common`  in this case) first before building `ProjectB/lambda1`. This is because
+maven is not able to find its way back up to the parent `ProjectB` to
+also build `ProjectB/common`. The challenge part here is to find the parent pom directory 
+especially for the projects with multiple level of submodules. The following implementation 
+assumes aws-lambda-builders have knowledge of the root package path for multimodule project.
+
+## Implementation
+
+### Build Workflow
+
+#### Step 1: Check Java version and emit warning
+
+Check whether the local JDK version is <= Java 8, and if it is not, emit a
+warning that the built artifact may not run in Lambda unless a) the project is
+properly configured (i.e. using `targetCompatibility`) or b) the project is
+built within a Lambda-compatibile environment like `lambci`.
+
+We use Maven to check the actual JVM version Maven is using in case it has been 
+configured to use a different one than can be found on the PATH.
+
+#### Step 2: Build and package
+
+We leverage Maven to do all the heavy lifting for executing the`mvn package` which
+will resolve and download the dependencies and build the project. Built java classes 
+will be located in `target/classes`. Then we use `mvn dependency:copy-dependenceis` to copy
```

Comment:
this `target` folder should not be within the `lambda` directory but instead within a scratch directory that you will get as part of the workflow. One of the design goals is to prevent the builder from writing to the Function's source as much as possible.

Reply:
I see. I was actually using `CopySourceAction` to copy the source code to the scratch directory in my first commit. Maven doesn't support change the target directory from command line, so we'd need to copy the whole project and it could take long time especially for large project. Will update.

Reply:
For multimodule project, builder also needs to know the root directory in order to copy the whole project.